Drop NA from leaf colour values before passing to np.unique()

Calling `np.unique()` on data from a pandas DataFrame which contains a mix of string and NaN values generates an error:

```
malariagen_data/anoph/distance.py:530: in plot_njt
    fig = anjl.plot(
../../../.cache/pypoetry/virtualenvs/malariagen-data-XR4NpaW--py3.12/lib/python3.12/site-packages/anjl/_plot.py:79: in plot
    normalise_color_params(
../../../.cache/pypoetry/virtualenvs/malariagen-data-XR4NpaW--py3.12/lib/python3.12/site-packages/anjl/_plot.py:198: in normalise_color_params
    unique_color_values = np.unique(leaf_color_values)
../../../.cache/pypoetry/virtualenvs/malariagen-data-XR4NpaW--py3.12/lib/python3.12/site-packages/numpy/lib/_arraysetops_impl.py:289: in unique
    ret = _unique1d(ar, return_index, return_inverse, return_counts,
...
E           TypeError: '<' not supported between instances of 'float' and 'str'
```

Possible solution is to drop NA values before passing to numpy.

--- a/anjl/_plot.py
+++ b/anjl/_plot.py
@@ -39,7 +39,8 @@ def plot(
     category_orders: params.category_order = None,
     leaf_legend: params.leaf_legend = True,
     edge_legend: params.edge_legend = False,
-    default_line_color: params.default_line_color = "black",
+    default_line_color: params.default_line_color = "gray",
+    na_color: params.na_color = "black",
     width: params.fig_width = 700,
     height: params.fig_height = 600,
     render_mode: params.render_mode = "auto",
@@ -64,6 +65,16 @@ def plot(
     if color is None and symbol is None:
         leaf_legend = edge_legend = False
 
+    # Handle missingness in a consistent way. This ensures that nodes and edges which
+    # have a missing value for either color or symbol column are still drawn in the
+    # plot.
+    if leaf_data is not None:
+        leaf_data = leaf_data.copy()
+        if color is not None:
+            leaf_data[color] = leaf_data[color].fillna("<NA>")
+        if symbol is not None:
+            leaf_data[symbol] = leaf_data[symbol].fillna("<NA>")
+
     # Decorate the tree.
     df_internal_nodes, df_leaf_nodes, df_edges = decorate_tree(
         Z=Z,
@@ -83,6 +94,7 @@ def plot(
             color_discrete_sequence=color_discrete_sequence,
             color_discrete_map=color_discrete_map,
             default_line_color=default_line_color,
+            na_color=na_color,
         )
     )
 
@@ -183,6 +195,7 @@ def normalise_color_params(
     color_discrete_sequence: params.color_discrete_sequence,
     color_discrete_map: params.color_discrete_map,
     default_line_color: params.default_line_color,
+    na_color: params.na_color,
 ) -> tuple[
     params.category_order, params.color_discrete_sequence, params.color_discrete_map
 ]:
@@ -191,11 +204,8 @@ def normalise_color_params(
         # and normalise the category_orders, color_discrete_sequence and
         # color_discrete_map parameters.
 
-        # Access the leaf colors.
-        leaf_color_values = np.asarray(leaf_data[color].dropna().values)
-
         # Find all unique color values.
-        unique_color_values = np.unique(leaf_color_values)
+        unique_color_values = leaf_data[color].unique()
 
         # Normalise category orders.
         if category_orders is None:
@@ -219,6 +229,9 @@ def normalise_color_params(
         # Set a default color for edges where descendant leaves have
         # different colors.
         color_discrete_map_prepped[""] = default_line_color
+
+        # Set a color for missing data.
+        color_discrete_map_prepped["<NA>"] = na_color
 
         color_discrete_map = color_discrete_map_prepped
 

--- a/anjl/_plot.py
+++ b/anjl/_plot.py
@@ -192,7 +192,7 @@ def normalise_color_params(
         # color_discrete_map parameters.
 
         # Access the leaf colors.
-        leaf_color_values = np.asarray(leaf_data[color].values)
+        leaf_color_values = np.asarray(leaf_data[color].dropna().values)
 
         # Find all unique color values.
         unique_color_values = np.unique(leaf_color_values)

--- a/anjl/params.py
+++ b/anjl/params.py
@@ -79,6 +79,8 @@ default_line_color: TypeAlias = Annotated[
     str, "Line color to use for edges where descendants are different colors."
 ]
 
+na_color: TypeAlias = Annotated[str, "Color to use where data are missing."]
+
 fig_width: TypeAlias = Annotated[
     int | float | None,
     "Figure width in pixels (px).",

--- a/sandbox/plot.ipynb
+++ b/sandbox/plot.ipynb
@@ -12,6 +12,7 @@
     "import zarr\n",
     "from scipy.spatial.distance import squareform\n",
     "import pandas as pd\n",
+    "import numpy as np\n",
     "\n",
     "plotly.io.renderers.default = \"plotly_mimetype+notebook_connected\""
    ]
@@ -139,7 +140,7 @@
     "        [\"A\", \"apple\", \"carrot\", \"pigeon\", \"Ferrari\"],\n",
     "        [\"B\", \"apple\", \"pea\", \"ladybird\", \"Lamborghini\"],\n",
     "        [\"C\", \"banana\", \"cabbage\", \"pigeon\", \"Landrover\"],\n",
-    "        [\"D\", \"mango\", \"cabbage\", \"spider\", \"Lamborghini\"],\n",
+    "        [\"D\", \"mango\", \"cabbage\", \"spider\", np.nan],\n",
     "    ],\n",
     "    columns=[\"name\", \"fruit\", \"vegetable\", \"animal\", \"car\"],\n",
     ")\n",
@@ -164,6 +165,16 @@
    "outputs": [],
    "source": [
     "anjl.paint_internal(Z1, ld1[\"fruit\"].values)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "0a2a712a-f025-474d-90a1-948322e3a694",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "anjl.paint_internal(Z1, ld1[\"car\"].values)"
    ]
   },
   {
@@ -200,6 +211,26 @@
     "    color=\"fruit\",\n",
     "    marker_size=20,\n",
     "    symbol=\"car\",\n",
+    "    internal_marker_size=5,\n",
+    "    line_width=4,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "4d9d6a48-691f-4aff-8832-3814d8639ac0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "anjl.plot(\n",
+    "    Z=Z1r,\n",
+    "    leaf_data=ld1,\n",
+    "    hover_name=\"name\",\n",
+    "    hover_data=[\"fruit\", \"vegetable\", \"animal\", \"car\"],\n",
+    "    color=\"car\",\n",
+    "    marker_size=20,\n",
+    "    symbol=\"fruit\",\n",
     "    internal_marker_size=5,\n",
     "    line_width=4,\n",
     ")"

--- a/tests/test_plot.py
+++ b/tests/test_plot.py
@@ -1,4 +1,5 @@
 import plotly.graph_objects as go
+import numpy as np
 import anjl
 
 
@@ -17,4 +18,21 @@ def test_wikipedia_example():
     D, _ = anjl.data.wikipedia_example()
     Z = anjl.canonical_nj(D)
     fig = anjl.plot(Z)
+    assert isinstance(fig, go.Figure)
+
+
+def test_mosquitoes():
+    D, leaf_data = anjl.data.mosquitoes()
+    Z = anjl.rapid_nj(D)
+    fig = anjl.plot(Z, leaf_data=leaf_data, color="location", symbol="taxon")
+    assert isinstance(fig, go.Figure)
+
+
+def test_gh38():
+    # https://github.com/alimanfoo/anjl/issues/38
+    D, leaf_data = anjl.data.mosquitoes()
+    leaf_data.loc[0, "taxon"] = np.nan
+    leaf_data.loc[1, "location"] = np.nan
+    Z = anjl.rapid_nj(D)
+    fig = anjl.plot(Z, leaf_data=leaf_data, color="location", symbol="taxon")
     assert isinstance(fig, go.Figure)

